### PR TITLE
Add `DuplicateImportsRule`

### PR DESCRIPTION
**Pull request by sammy-SC**
implements #1881 

Any recommendations for improvement are very welcome.

if you are happy with implementation, I will be happy to get started on autocorrection.

### CODE REVIEW

**Comment by ornithocoder** [comment]
Hi @sammy-SC. I don't know useful my suggestion would be for the implementation. But an easy way to find dups is using a `Set`. Inserting into a set - using `func insert(_:)` - returns a tuple with `inserted` and `newMember`. So if it fails trying to insert into the set, it's because the value is already there, therefore a dup violation.

**Comment by sammy-SC** [comment]
Hello @ornithocoder 
thank you very much for the suggestion.

I was considering that approach however I couldn't find a way to deal with nested modules.
For example you can have following code
```
import A.B
import A
```
and this should throw a warning on first line as that module has already been imported.

Hope it makes sense and explains why I chose to go with comparing lines among themselves with two nested for loops.

File: Source/SwiftLintFramework/Rules/DuplicateImportsRule.swift [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    public init() {}
+
+    public static let description = RuleDescription(
+        identifier: "duplicate_imports",
+        name: "Duplicate Imports",
+        description: "Imports should be unique.",
+        kind: .style,
+        nonTriggeringExamples: [
+            "import A\nimport B\nimport C",
+            "import A.B\nimport import A.C"
```
Comment (ornithocoder) [inline Source/SwiftLintFramework/Rules/DuplicateImportsRule.swift]: There's a double `import` for `A.C`.

**Comment by ornithocoder** [comment]
Hi @sammy-SC. Regarding the `#ifdef/else/endif`, it doesn't show on the structure

```json
{
  "key.diagnostic_stage" : "source.diagnostic.stage.swift.parse",
  "key.substructure" : [

  ],
  "key.offset" : 0,
  "key.length" : 67
}
```

But "shows" on the syntax map.

```json
[
  {
    "type" : "source.lang.swift.syntaxtype.buildconfig.keyword",
    "offset" : 0,
    "length" : 3
  },
  {
    "type" : "source.lang.swift.syntaxtype.buildconfig.id",
    "offset" : 4,
    "length" : 5
  },
  {
    "type" : "source.lang.swift.syntaxtype.attribute.builtin",
    "offset" : 14,
    "length" : 9
  },
  {
    "type" : "source.lang.swift.syntaxtype.keyword",
    "offset" : 24,
    "length" : 6
  },
  {
    "type" : "source.lang.swift.syntaxtype.identifier",
    "offset" : 31,
    "length" : 5
  },
  {
    "type" : "source.lang.swift.syntaxtype.buildconfig.keyword",
    "offset" : 37,
    "length" : 5
  },
  {
    "type" : "source.lang.swift.syntaxtype.keyword",
    "offset" : 47,
    "length" : 6
  },
  {
    "type" : "source.lang.swift.syntaxtype.identifier",
    "offset" : 54,
    "length" : 5
  },
  {
    "type" : "source.lang.swift.syntaxtype.buildconfig.keyword",
    "offset" : 60,
    "length" : 6
  }
]
```

But honestly, I don't know what to do in this situation. Probably ignore duplicates if they're wrapped by flags? Specially if the `#if/endif` has `elif`s.

File: Source/SwiftLintFramework/Rules/DuplicateImportsRule.swift (hunk 1 above)
Comment (sammy-SC) [reply Source/SwiftLintFramework/Rules/DuplicateImportsRule.swift]: @ornithocoder thank you, I've fixed that.

**Comment by sammy-SC** [comment]
Hello @ornithocoder

I've done what you suggested, ignoring imports in conditional compilation directives.

**Comment by marcelofabri** [comment]
I'm curious on why https://github.com/apple/swift/blob/d01c9fa136bb35b801acb948ab2ee4f1d6cb3e9c/stdlib/public/Platform/msvcrt.swift#L14:1 triggered the warning 🤔

**Comment by sammy-SC** [comment]
@marcelofabri I'm sorry I must have missed that.
I've fixed the issue and made sure all of the warnings are correct.

**Comment by marcelofabri** [comment]
Hmm, apparently this rule made linting way slower on oss-check:

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          12 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Linting Aerial with this PR took 0.73s vs 0.42s on master (73% slower)</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Linting Alamofire with this PR took 5.43s vs 3.49s on master (55% slower)</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Linting Firefox with this PR took 22.14s vs 14.81s on master (49% slower)</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Linting Kickstarter with this PR took 31.66s vs 24.36s on master (29% slower)</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Linting Moya with this PR took 2.82s vs 2.49s on master (13% slower)</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Linting Nimble with this PR took 2.98s vs 2.52s on master (18% slower)</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Linting Quick with this PR took 0.85s vs 0.79s on master (7% slower)</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Linting Realm with this PR took 4.95s vs 4.28s on master (15% slower)</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Linting SourceKitten with this PR took 1.52s vs 1.34s on master (13% slower)</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Linting Sourcery with this PR took 6.7s vs 6.71s on master (0% faster)</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Linting Swift with this PR took 19.88s vs 21.21s on master (6% faster)</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Linting WordPress with this PR took 17.32s vs 22.63s on master (23% faster)</td>
    </tr>
  </tbody>
</table>


I'm re-running it just to be sure it wasn't caused by a Circle CI instability.

**Comment by sammy-SC** [comment]
@marcelofabri are the new results sufficient or should I look into improving performance?

**Comment by marcelofabri** [comment]
@sammy-SC I think it's fine now. I'll try to review this in the following days.

**Comment by sammy-SC** [comment]
@marcelofabri could you please check this PR when you have time? 

I think it would be a useful addition to SwiftLint.

**Comment by marcelofabri** [comment]
@sammy-SC Sorry for the delay. 

I think we should make sure we're handling all the import forms, including the less-known import kind (https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/Declarations.html#//apple_ref/doc/uid/TP40014097-CH34-ID354). 

Or we can skip that for now and make the rule opt-in.

File: Source/SwiftLintFramework/Rules/DuplicateImportsRule.swift [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+//  SwiftLint
+//
+//  Created by Samuel Susla on 1/13/18.
+//  Copyright © 2018 Realm. All rights reserved.
+//
+
+import Foundation
+import SourceKittenFramework
+
+extension Line {
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/DuplicateImportsRule.swift]: this extension should be `private` (and I would move it to the end of this file)

File: Source/SwiftLintFramework/Rules/DuplicateImportsRule.swift [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+public struct DuplicateImportsRule: ConfigurationProviderRule {
+    public var configuration = SeverityConfiguration(.warning)
+
+    public init() {}
+
+    public static let description = RuleDescription(
+        identifier: "duplicate_imports",
+        name: "Duplicate Imports",
+        description: "Imports should be unique.",
+        kind: .style,
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/DuplicateImportsRule.swift]: `.idiomatic`?

File: Source/SwiftLintFramework/Rules/DuplicateImportsRule.swift [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        let start = "#if"
+        let end = "#endif"
+        let contents = file.contents.bridge()
+
+        let ranges = file.syntaxMap.tokens
+            .filter { token -> Bool in
+                guard let kind = SyntaxKind(rawValue: token.type) else {
+                    return false
+                }
+                return kind == SyntaxKind.buildconfigKeyword
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/DuplicateImportsRule.swift]: you can use just `.buildconfigKeyword`

File: Source/SwiftLintFramework/Models/MasterRuleList.swift [hunk 5]
```diff
@@ -5,6 +5,7 @@
 //  Created by Scott Hoyt on 12/28/15.
 //  Copyright © 2015 Realm. All rights reserved.
 //
+
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Models/MasterRuleList.swift]: 🤔

Did you manually add this line?

File: Source/SwiftLintFramework/Rules/DuplicateImportsRule.swift (hunk 2 above)
Comment (sammy-SC) [reply Source/SwiftLintFramework/Rules/DuplicateImportsRule.swift]: Done

File: Source/SwiftLintFramework/Rules/DuplicateImportsRule.swift (hunk 3 above)
Comment (sammy-SC) [reply Source/SwiftLintFramework/Rules/DuplicateImportsRule.swift]: Changed.

File: Source/SwiftLintFramework/Rules/DuplicateImportsRule.swift (hunk 4 above)
Comment (sammy-SC) [reply Source/SwiftLintFramework/Rules/DuplicateImportsRule.swift]: Changed

File: Source/SwiftLintFramework/Models/MasterRuleList.swift (hunk 5 above)
Comment (sammy-SC) [reply Source/SwiftLintFramework/Models/MasterRuleList.swift]: I'm sorry, I removed the line

**Comment by sammy-SC** [comment]
@marcelofabri 
I know it has been a while but I managed to change it so all import forms are accounted for.

Please when you have a time could you take another look at this? 

Thank you

**Comment by sammy-SC** [comment]
@marcelofabri could you please rerun the speed check?

**Comment by sammy-SC** [comment]
@marcelofabri I resolved merge conflicts. 
This rule seems to slow down SwiftLint, do we want to do anything about it? If so do you happen to have any tips/tricks on how to go about it?

**Comment by marcelofabri** [comment]
@sammy-SC I ran oss-check locally and it seems it's not too bad:

```
Message: Linting Aerial with this PR took 0.32s vs 0.3s on master (6% slower)
Message: Linting Alamofire with this PR took 2.23s vs 2.19s on master (1% slower)
Message: Linting Firefox with this PR took 9.42s vs 9.38s on master (0% slower)
Message: Linting Kickstarter with this PR took 13.03s vs 12.43s on master (4% slower)
Message: Linting Moya with this PR took 1.39s vs 1.38s on master (0% slower)
Message: Linting Nimble with this PR took 1.12s vs 1.14s on master (1% faster)
Message: Linting Quick with this PR took 0.36s vs 0.35s on master (2% slower)
Message: Linting Realm with this PR took 2.13s vs 2.13s on master (0% slower)
Message: Linting SourceKitten with this PR took 0.76s vs 0.79s on master (3% faster)
Message: Linting Sourcery with this PR took 3.5s vs 3.51s on master (0% faster)
Message: Linting Swift with this PR took 22.37s vs 21.82s on master (2% slower)
Message: Linting WordPress with this PR took 12.17s vs 11.49s on master (5% slower)
```

**Comment by sammy-SC** [comment]
@marcelofabri please let me know if there is anything else you would like me to change. Any tips on improving the code are very welcome, even if it was to improve readability / clarify of the solution.

If approved I would be more than happy to add autocorrection.

**Comment by jpsim** [comment]
OSSCheck highlighted a number of false positives. For example, `Realm` and `Realm.Private` are different modules, but this is triggering a violation:

```swift
import Realm
import Realm.Private
```

**Comment by sammy-SC** [comment]
@jpsim that's intentional, is that not correct?

The idea was that if you have imports such as 

```
import Foundation
import Foundation.Date
```

would get highlighted as duplicate import.

Is that not desired? 

If not, I'm happy to change it.

**Comment by jpsim** [comment]
Thank you for being so patient on this one @sammy-SC! I finally had some time to update this, refactor it a little and came to a decision on keeping the rule more powerful but with a few rare false positives.

**Comment by sammy-SC** [comment]
@jpsim thanks for the help. I will check out the refactoring. 

It took a little over a year so not too bad :)

**Pull request merged**